### PR TITLE
resolve_composes: add support for modular_koji_tags

### PR DESCRIPTION
**Pull request by mlangsdorf**
fixes osbs6833

Add support to pre_resolve_composes to parse modular_koji_tags and pass that on to the ODCS engine.

Signed-off-by: Mark Langsdorf <mlangsdo@redhat.com>

### CODE REVIEW

**Comment by mlangsdorf** [comment]
just realized we don't have a schema for pre_resolve_compose's data files.  Not sure if that can be fixed, either.

File: atomic_reactor/schemas/container.json [hunk 1]
```diff
@@ -82,9 +82,42 @@
...
+              "enum": ["none", "iso", "runtime", "devel", "all"]
+          }
+        },
+        "multilib_arches": {
+          "description": "tags of modular Koji content to include in the build",
+          "type": ["array", "null"],
+          "items": {
+              "type": "string"
+          }
+        },
```
Comment (twaugh) [inline atomic_reactor/schemas/container.json]: Given that we don't do anything with these properties, do we really want to allow them in container.yaml?

Comment (mlangsdorf) [reply atomic_reactor/schemas/container.json]: we support multlib_arches and multilib_methods.  See #1084.

Fixed the description, though.

Comment (twaugh) [reply atomic_reactor/schemas/container.json]: Ah, thanks. and I see include_unpublished_pulp_repos is also used.

What about no_deps and check_deps?

Comment (mlangsdorf) [reply atomic_reactor/schemas/container.json]: They're options for the flags argument same as include_unpublished_pulp_repos and use the same syntax.

File: atomic_reactor/plugins/pre_resolve_composes.py [hunk 2]
```diff
@@ -431,6 +432,8 @@ def render_modules_request(self):
         }
         if self.arches:
             request['arches'] = self.arches
+        if self.modular_tags:
+            request['modular_koji_tags'] = ' '.join(self.modular_tags)
```
Comment (athos-ribeiro) [inline atomic_reactor/plugins/pre_resolve_composes.py]: In the previous commit, `start_compose` docstring says it is supposed to receive a list of strings (ODCS code also seems to be expecting a list). Is there a reason for having a string here?

Comment (mlangsdorf) [reply atomic_reactor/plugins/pre_resolve_composes.py]: I got confused. Fixed.

File: atomic_reactor/schemas/container.json (hunk 1 above)
Comment (rcerven) [reply atomic_reactor/schemas/container.json]: but I don't see those 2 options used anywhere

**Comment by rcerven** [comment]
code looks good

File: atomic_reactor/schemas/container.json (hunk 1 above)
Comment (mlangsdorf) [reply atomic_reactor/schemas/container.json]: pre_resolve_composes scoops up everything that isn't used somewhere else into flags and passed them to ODCS as a flag.

**Comment by athos-ribeiro** [comment]
LGTM

About the test suite:

one of the parameters passed to `test_valid_source_config` at `tests/test_source.py:190` is not valid anymore. The extra_key entry could be moved to some on the invalid tests and new entries with valid examples could be place in `test_valid_source_config`

File: atomic_reactor/schemas/container.json (hunk 1 above)
Comment (rcerven) [reply atomic_reactor/schemas/container.json]: this is only place where any flags are set

https://github.com/projectatomic/atomic-reactor/blob/master/atomic_reactor/plugins/pre_resolve_composes.py#L378

**Comment by mlangsdorf** [comment]
Still can't figure out why flatpak tests are failing.  I can't reproduce locally.

File: tests/plugins/test_flatpak_create_oci.py [hunk 3]
```diff
@@ -475,7 +475,7 @@ def build_export(repo, directory, branch):
       '--oci', '@filename', '@name', '@branch'],
      MockFlatpak.build_bundle),
     # For build-export, we assume the latest flatpak version for flatpak_module_tools compatibility
-    (['flatpak', 'build-export', '@repo', '@directory', '@branch', '--disable-sandbox'],
```
Comment (athos-ribeiro) [inline tests/plugins/test_flatpak_create_oci.py]: I did reproduce the failure locally and reverting this change will fix the broken test.

See discussion #1160 as reference for why we need this flag.

About reproducing the error locally: It seems this tests are being skipped even using the same script as the CI does. I needed to install krb5-devel and python-koji in the container to get it to run. Maybe some package got updated in Fedora and didn't reach CI mirrors yet?

Comment (mlangsdorf) [reply tests/plugins/test_flatpak_create_oci.py]: I added that change to get my local tests to work.  That is so crazy.  Well, I deleted that commit and things are green again.

File: tests/test_odcs_util.py [hunk 4]
```diff
@@ -82,13 +82,16 @@ def compose_json(state, state_name, source_type='module', source=MODULE_NSV,
...
-    ('my-tag', 'tag', None, [], None),
-    ('my-tag', 'tag', ['spam', 'bacon', 'eggs'], ['spam', 'bacon', 'eggs'], None),
-    ('my-tag', 'tag', ['spam', 'bacon', 'eggs'], ['spam', 'bacon', 'eggs'], ['B456', 'R123']),
-    ('my-tag', 'tag', ['spam', 'bacon', 'eggs'], ['spam', 'bacon', 'eggs'], ""),
-    ('my-tag', 'tag', ['spam', 'bacon', 'eggs'], ['spam', 'bacon', 'eggs'], []),
+@pytest.mark.parametrize(('source', 'source_type', 'packages', 'expected_packages', 'sigkeys',
+                          'modular_koji_tags', 'expected_koji_tags'), (
+    (MODULE_NSV, 'module', None, None, None, [], None),
+    (MODULE_NSV, 'module', None, None, None, ['release'], ['release']),
+    ('my-tag', 'tag', None, [], None, 'release', None),
```
Comment (twaugh) [inline tests/test_odcs_util.py]: Should that be `['release']`?

File: tests/test_odcs_util.py [hunk 5]
```diff
@@ -82,13 +82,16 @@ def compose_json(state, state_name, source_type='module', source=MODULE_NSV,
...
+    ('my-tag', 'tag', ['spam', 'bacon', 'eggs'], ['spam', 'bacon', 'eggs'], None, None, None),
+    ('my-tag', 'tag', ['spam', 'bacon', 'eggs'], ['spam', 'bacon', 'eggs'], ['B456', 'R123'],
+     'release', None),
+    ('my-tag', 'tag', ['spam', 'bacon', 'eggs'], ['spam', 'bacon', 'eggs'], "", None, None),
+    ('my-tag', 'tag', ['spam', 'bacon', 'eggs'], ['spam', 'bacon', 'eggs'], [], 'release', None),
```
Comment (twaugh) [inline tests/test_odcs_util.py]: Should that be `['release']`?

File: tests/test_odcs_util.py (hunk 4 above)
Comment (mlangsdorf) [reply tests/test_odcs_util.py]: No, modular koji tags are not a valid parameter for 'tag' releases and get ignored.

File: tests/test_odcs_util.py (hunk 5 above)
Comment (mlangsdorf) [reply tests/test_odcs_util.py]: No, modular koji tags are not a valid parameter for 'tag' releases and get ignored.

Comment (twaugh) [reply tests/test_odcs_util.py]: Yes, I know we expect `None` for the final value there, but the value before that is a string not a list. I don't think that would be allowed by the schema, so I don't think it's testing something the logic would actually see.

Comment (mlangsdorf) [reply tests/test_odcs_util.py]: Ah, okay.  Yes, you're right.  Fixed.

File: atomic_reactor/schemas/container.json (hunk 1 above)
Comment (twaugh) [reply atomic_reactor/schemas/container.json]: What's the resolution to this?

Comment (mlangsdorf) [reply atomic_reactor/schemas/container.json]: They're valid flags, I'd like to keep them in the schema even if we don't support them in pre_resolve_composes.

Comment (mlangsdorf) [reply atomic_reactor/schemas/container.json]: Removed the flags we don't support.

**Comment by mlangsdorf** [comment]
Added to release notes under improvements
"- Add support for 'modular_koji_tags' in ODCS compose builds"

**Pull request merged**